### PR TITLE
refactor!: compare with origin bytes during the transactions

### LINKED ISSUE #2198: Using the remote bytes the instead of the serialized bytes during the transaction
### What type of enhancement is this?

Refactor, Tech debt reduction

### What does the enhancement do?

```rust
  /// Builds a update table info transaction, it expected the remote value equals the `current_current_table_info_value`.
    /// It retrieves the latest value if the comparing failed.
    pub(crate) fn build_update_txn(
        &self,
        table_id: TableId,
        current_table_info_value: &TableInfoValue,
        new_table_info_value: &TableInfoValue,
    ) -> Result<(
        Txn,
        impl FnOnce(&Vec<TxnOpResponse>) -> Result<Option<TableInfoValue>>,
    )> {
        let key = TableInfoKey::new(table_id);
        let raw_key = key.as_raw_key();
        let raw_value = current_table_info_value.try_as_raw_value()?;

        let txn = Txn::new()
            .when(vec![Compare::with_value(
                raw_key.clone(),
                CompareOp::Equal,
                raw_value.clone(),
            )])
            .and_then(vec![TxnOp::Put(
                raw_key.clone(),
                new_table_info_value.try_as_raw_value()?,
            )])
            .or_else(vec![TxnOp::Get(raw_key.clone())]);
```

We retrieve the `remote value` from etcd, deserialize it into the `TableInfoValue`, and build the transaction with the `TableInfoValue` serialization value. However, we must ensure it during the serialization, the `remote value` without changes. This way is problematic, and it's better to build the transaction with the `raw value` directly.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. Wrap values with `DeserializedValueWithBytes` struct.
2. Compare with origin bytes during the transactions

This PR modifies the persistent structs used in meta's procedures, ensuring all old procedures are finished before starting a new version of metasrv.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
close #2198

### CODE REVIEW

File: src/common/meta/src/key.rs [hunk 1]
```diff
@@ -570,11 +686,39 @@ mod tests {
...
+    #[test]
+    fn test_deserialized_value_with_bytes() {
+        let region_route = new_test_region_route();
+        let region_routes = vec![region_route.clone()];
+
+        let expected_region_routes =
+            TableRouteValue::new(vec![region_route.clone(), region_route.clone()]);
+        let expected = serde_json::to_vec(&expected_region_routes).unwrap();
+
+        //Serialize behaviors:
```
Comment (fengjiachun) [inline src/common/meta/src/key.rs:703]:
```suggestion
        // Serialize behaviors:
```

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/key/table_info.rs [hunk 2]
```diff
@@ -143,15 +143,15 @@ impl TableInfoManager {
...
         new_table_info_value: &TableInfoValue,
     ) -> Result<(
         Txn,
-        impl FnOnce(&Vec<TxnOpResponse>) -> Result<Option<TableInfoValue>>,
+        impl FnOnce(&Vec<TxnOpResponse>) -> Result<Option<DeserializedValueWithBytes<TableInfoValue>>>,
     )> {
         let key = TableInfoKey::new(table_id);
         let raw_key = key.as_raw_key();
-        let raw_value = current_table_info_value.try_as_raw_value()?;
+        let raw_value = current_table_info_value.bytes.to_vec();
```
Comment (MichaelScofield) [inline src/common/meta/src/key/table_info.rs:154]: Not exposing `bytes` field and putting `.bytes.to_vec()` in a method brings better encapsulation.

File: src/common/meta/src/key.rs [hunk 3]
```diff
@@ -336,7 +446,7 @@ impl TableMetadataManager {
             .build_delete_txn(table_id, table_info_value)?;
 
         // Deletes datanode table key value pairs.
-        let distribution = region_distribution(&table_route_value.region_routes)?;
+        let distribution = region_distribution(&table_route_value.inner.region_routes)?;
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:449]: We can't. Some places still need both `bytes` and `inner`.

File: src/common/meta/src/key.rs [hunk 4]
```diff
@@ -155,6 +159,107 @@ macro_rules! ensure_values {
...
+
+        Ok(value)
+    }
+}
+
+impl<T: Serialize + DeserializeOwned + Clone> Clone for DeserializedValueWithBytes<T> {
+    fn clone(&self) -> Self {
+        Self {
+            bytes: self.bytes.clone(),
+            inner: self.inner.clone(),
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:224]: I see, But I think it's better to minimize `T`.

File: src/common/meta/src/key.rs [hunk 5]
```diff
@@ -155,6 +159,107 @@ macro_rules! ensure_values {
...
+    /// - Serialize behaviors:
+    ///
+    /// The `inner` field will be ignored.
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        // Safety: The original bytes are always JSON encoded.
+        // It's more efficiently than `serialize_bytes`.
+        serializer.serialize_str(&String::from_utf8_lossy(&self.bytes))
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:198]: [serialize_bytes](https://docs.rs/serde/1.0.188/serde/trait.Serializer.html#tymethod.serialize_bytes) suffers from the same problem. For JSON format, `serialize_bytes` always treats the u8 array as the u32 array.

**Comment by WenyXu** [comment]
@MichaelScofield PTAL

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**